packages/discovery-provider/src/queries/get_extended_purchase_gate.py

```diff
@@ -0,0 +1,240 @@
+from src.utils.session_manager import SessionManager
+
+logger = logging.getLogger(__name__)
+
+
+class GetPurchaseInfoArgs(TypedDict):
+    content_id: int
+    content_type: str
+    content: Union[Track, Playlist]
+
+
+class Split(TypedDict):
+    user_id: int
+    percentage: float
+
+
+class USDCPurchaseCondition(TypedDict):
+    price: int
+    splits: List[Split]
+
+
+class PurchaseGate(TypedDict):
+    usdc_purchase: USDCPurchaseCondition
+
+
+class ExtendedSplit(Split):
+    amount: int
+    payout_wallet: str
+    eth_wallet: str
+
+
+class ExtendedUSDCPurchaseCondition(TypedDict):
+    price: int
+    splits: List[ExtendedSplit]
+
+
+class ExtendedPurchaseGate(TypedDict):
+    usdc_purchase: ExtendedUSDCPurchaseCondition
+
+
+class LegacyUSDCPurchaseCondition(TypedDict):
+    price: int
+    splits: Dict[str, int]
+
+
+class LegacyPurchaseGate(TypedDict):
+    usdc_purchase: LegacyUSDCPurchaseCondition
+
+
+class FollowGate(TypedDict):
+    follow_user_id: int
+
+
+class TipGate(TypedDict):
+    tip_user_id: int
+
+
+class NFTCollection(TypedDict):
+    chain: str
+    address: str
+    name: str
+    imageUrl: Optional[str]
+    externalLink: Optional[str]
+
+
+class NFTGate(TypedDict):
+    nft_collection: NFTCollection
+
+
+AccessGate = Union[PurchaseGate, FollowGate, TipGate, NFTGate]
+
+
+# Allow up to 6 decimal points for split percentages
+percentage_decimals = 6
+percentage_multiplier = 10**percentage_decimals
+
+
+cents_to_usdc_multiplier = 10**4
+
+
+def calculate_split_amounts(price: int, splits: List[Split]):
+    """
+    Deterministically calculates the USDC amounts to pay to each person,
+    adjusting for rounding errors and ensuring the total matches the price.
+    """
+    price_in_usdc = int(cents_to_usdc_multiplier * price)
+    running_total = 0
+    new_splits: List[Dict] = []
+    for index in range(len(splits)):
+        split = splits[index]
+        # multiply percentage to make it a whole number
+        percentage_whole = int(split["percentage"] * percentage_multiplier)
+        # do safe integer math on the price
+        amount = percentage_whole * price_in_usdc
+        # divide by the percentage multiplier afterward, and convert percent
+        amount = amount / (percentage_multiplier * 100)
+        # round towards zero, it'll round up later as necessary
+        amount_in_usdc = int(amount)
+        new_split: Dict = cast(dict, split)
+        new_split["amount"] = amount_in_usdc
+        # save the fractional component and index for rounding/sorting later
+        new_split["_amount_fractional"] = amount - amount_in_usdc
+        new_split["_index"] = index
+        new_splits.append(new_split)
+        running_total += amount_in_usdc
+    # Resolve rounding errors by iteratively choosing the highest fractional
+    # rounding errors to round up until the running total is correct
+    new_splits.sort(key=lambda item: (-item["_amount_fractional"], item["amount"]))
```

Comment:
Is this following a specific algorithm used for resolving rounding errors? Feels a bit like voodoo, but if it's an implementation of a known solve I'm less scared.

Reply:
it is voodoo